ffi/memorymanager.cpp

```diff
@@ -0,0 +1,334 @@
+    unsigned Alignment) {
+    if (!Alignment)
+        Alignment = 16;
+
+    assert(!(Alignment & (Alignment - 1)) &&
+           "Alignment must be a power of two.");
+
+    uintptr_t RequiredSize =
+        Alignment * ((Size + Alignment - 1) / Alignment + 1);
+    uintptr_t Addr = 0;
+
+    LLVM_DEBUG(dbgs() << "Allocating " << format_hex(RequiredSize, 2, true)
+                      << " bytes for ");
+
+    MemoryGroup &MemGroup = [&]() -> MemoryGroup & {
+        switch (Purpose) {
+        case AllocationPurpose::Code:
+            LLVM_DEBUG(dbgs() << "CodeMem at ");
+            return CodeMem;
+        case AllocationPurpose::ROData:
+            LLVM_DEBUG(dbgs() << "RODataMem at ");
+            return RODataMem;
+        case AllocationPurpose::RWData:
+            LLVM_DEBUG(dbgs() << "RWDataMem at ");
+            return RWDataMem;
+        }
+        llvm_unreachable("Unknown LlvmliteMemoryManager::AllocationPurpose");
+    }();
+
+    // Look in the list of free memory regions and use a block there if one
+    // is available.
+    for (FreeMemBlock &FreeMB : MemGroup.FreeMem) {
+        if (FreeMB.Free.allocatedSize() >= RequiredSize) {
+            Addr = (uintptr_t)FreeMB.Free.base();
+            uintptr_t EndOfBlock = Addr + FreeMB.Free.allocatedSize();
+            // Align the address.
+            Addr = (Addr + Alignment - 1) & ~(uintptr_t)(Alignment - 1);
+
+            if (FreeMB.PendingPrefixIndex == (unsigned)-1) {
+                // The part of the block we're giving out to the user is now
+                // pending
+                MemGroup.PendingMem.push_back(
+                    sys::MemoryBlock((void *)Addr, Size));
+
+                // Remember this pending block, such that future allocations can
+                // just modify it rather than creating a new one
+                FreeMB.PendingPrefixIndex = MemGroup.PendingMem.size() - 1;
+            } else {
+                sys::MemoryBlock &PendingMB =
+                    MemGroup.PendingMem[FreeMB.PendingPrefixIndex];
+                PendingMB =
+                    sys::MemoryBlock(PendingMB.base(),
+                                     Addr + Size - (uintptr_t)PendingMB.base());
+            }
+
+            // Remember how much free space is now left in this block
+            FreeMB.Free = sys::MemoryBlock((void *)(Addr + Size),
+                                           EndOfBlock - Addr - Size);
+            LLVM_DEBUG(dbgs() << format_hex(Addr, 18, true) << "\n");
+            return (uint8_t *)Addr;
+        }
+    }
+
+    assert(false && "All memory must be pre-allocated");
+}
+
+static uintptr_t requiredPageSize(uintptr_t Size, uint32_t Alignment) {
+    static const size_t PageSize = sys::Process::getPageSizeEstimate();
+    // Use the same calculation as allocateSection because we need to be able to
+    // satisfy it.
+    uintptr_t RequiredSize =
+        Alignment * ((Size + Alignment - 1) / Alignment + 1);
```

Comment:
This uses the same oversizing that's used in allocateSection (which is done to let it align the address as well), but it can only do it on the full reservation. I was hoping the caller would ensure they're calculation had sufficient buffer for multiple calls to allocateDataSection/allocateCodeSection, but it looks like there may be circumstances where that's not true.

Reply:
Ah, thanks for looking at this PR! Tracing through things a bit I just noticed https://github.com/numba/llvmlite/pull/1009/files#r1394577484

Reply:
I think the reason for the discrepancy is that the reservation is made with the alignment of the code section's alignment (4), but the actual allocation is made with the max of the alignment of the code section (4) and the stub alignment (which is 8 on AArch64 on macOS, and maybe other platforms). So I think the code segment preallocation needs to be aligned to max(code section alignment, stub alignment) too.